### ISSUE TITLE
Improve support for arbitrary functions accepting `TypeForm` arguments

### ISSUE THREAD

**Issue by Daverball** [astral-sh/ruff#26625]
With the acceptance and implementation of PEP 747 in mypy and being no longer locked behind a feature-flag as of mypy 2.2 we can expect usage of `TypeForm` to increase, which includes functions that don't necessarily use the argument at runtime like `typing.cast`.

As a result there will be custom functions like that accept stringified forward references that ruff will not treat as forward references, so we will see false positives for rules like F401 and TC001-004.

Similar to the `extend-generics` setting, we may want to add something like a `extend-type-form-callables` setting which accepts a mapping of function names, to a list of argument names/positions, so we can treat the relevant subexpressions as typing expressions.

E.g. for `typing.cast` the configuration could look like this:

```toml
[lint.extend-type-form-callables]
"typing.cast" = [0, "typ"]
```

Meaning the first argument positionally and the `typ` keyword argument will be interpreted as a type expression on any call that resolves to the configured callable name, i.e. in this case `typing.cast`.

Eventually we could automate this with multi-file analysis by inspecting the parameter annotations of the called function, but for now a custom setting should be good enough.

**Comment by ronits2407** [comment]
Hi! I'd love to take this on as my first contribution.

I plan to implement this by:
1. Adding a new `extend-type-form-callables` field to `LintOptions` in `crates/ruff_workspace/src/options.rs`.
2. Creating an untagged enum `CallArgument` (to support both integer positions and string keyword arguments).
3. Piping this configuration down to `analyze/typing.rs` or wherever `typing.cast` and type expressions are currently checked to ensure these arguments are treated as type expressions instead of standard strings.

Let me know if this sounds like the right approach or if you have any pointers!

**Comment by MichaReiser** [comment]
Seems reasonable and with a ver clear evolution. The syntax feels a bit cryptic, I'm leaning towards `"typing.cast" = { position = 0, name = "typ" }`. We may also need to think about how to express positional-only and keyword argument only arguments (an entry must have either a `position` or `name` or both?)

**Comment by ronits2407** [comment]
> Seems reasonable and with a ver clear evolution. The syntax feels a bit cryptic, I'm leaning towards `"typing.cast" = { position = 0, name = "typ" }`. We may also need to think about how to express positional-only and keyword argument only arguments (an entry must have either a `position` or `name` or both?)

I completely agree that { position = 0, name = "typ" } is much clearer than mixing integers and strings in a list.

To handle positional-only vs keyword-only, we could just make position and name both optional fields in the struct. If only position is provided, it applies to positional usage; if only name is provided, it applies to keyword usage. If both are provided, it matches either. Does that sound like the right level of strictness?

I went ahead and pushed the above changes with the original untagged approach so you can see where the AST hooks are placed, but I'm more than happy to refactor the configuration schema to use this explicit { position, name } format instead. Let me know if that sounds good!

**Comment by MichaReiser** [comment]
> To handle positional-only vs keyword-only, we could just make position and name both optional fields in the struct. If only position is provided, it applies to positional usage; if only name is provided, it applies to keyword usage. If both are provided, it matches either. Does that sound like the right level of strictness?

That sounds about right. We may want to write our own schema to express the: it requires at least one or both requirement.

**Comment by ronits2407** [comment]
alright i'm working on that

**Comment by Daverball** [comment]
That sounds reasonable, just wanted to add that you'd still want a list of argument definitions, in case more than one argument of a function accepts a `TypeForm`.

I.e. `"typing.cast" = [{ position = 0, name = "typ" }]` rather than `"typing.cast" = { position = 0, name = "typ" }`

**Comment by ronits2407** [comment]
I've refactored the configuration schema to use the [{ position, name }] struct format as discussed! I also added some unit tests using the snapshot framework and verified that AST types are correctly mapped for the configured callables. The PR description has been updated and it's marked as ready for review again. Let me know if there's anything else needed!
